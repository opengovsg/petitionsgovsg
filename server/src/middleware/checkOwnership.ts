import { NextFunction, Request, Response } from 'express'
import { StatusCodes } from 'http-status-codes'
import { Signature, Post } from '~shared/types/base'
import { Message } from '../types/message-type'
import { ControllerHandler } from '../types/response-handler'
import { ModelDef } from '../types/sequelize'

type SignatureWithRelations = Signature & {
  userId: number
  post: Post
}
export type OwnershipCheck = ControllerHandler<
  { id: string },
  Message,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  any
>
export const checkOwnershipUsing = ({
  Signature,
  Post,
}: {
  Signature: ModelDef<Signature>
  Post: ModelDef<Post>
}): OwnershipCheck => {
  const checkOwnership: OwnershipCheck = async (
    req: Request,
    res: Response,
    next: NextFunction,
  ): Promise<Response | void> => {
    //TODO

    // const results = (await Signature.findOne({
    //   where: { id: req.params.id },
    //   attributes: ['userId'],
    //   include: [
    //     {
    //       model: Post,
    //     },
    //   ],
    // })) as unknown as SignatureWithRelations
    // if (!results) {
    //   return res
    //     .status(StatusCodes.BAD_REQUEST)
    //     .json({ message: 'No answer found with this ID' })
    // }
    // if (!req.user) {
    //   return res
    //     .status(StatusCodes.UNAUTHORIZED)
    //     .json({ message: 'User not signed in' })
    // }

    // const user = await User.findByPk(req.user.id)

    // if (!user) {
    //   return res.status(StatusCodes.FORBIDDEN).json({
    //     message: `User ${req.user.id} does not exist`,
    //   })
    // }

    next()
  }
  return checkOwnership
}
