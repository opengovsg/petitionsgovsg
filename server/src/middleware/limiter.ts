import rateLimit from 'express-rate-limit'

// Set limiter on authenticated routes to reduce burden on servers
export const limiter = rateLimit({
  windowMs: 1 * 60 * 1000, // 1 minute
  max: 10, // limit each IP to 60 requests per windowMs
  handler: function (req, res) {
    return res.status(429).json({
      error: 'You sent too many requests. Please wait a while then try again',
    })
  },
})
