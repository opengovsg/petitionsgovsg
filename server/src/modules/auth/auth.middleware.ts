import { StatusCodes } from 'http-status-codes'
import { ControllerHandler } from '../../types/response-handler'
import { decodeUserJWT } from '../../util/jwt'
export class AuthMiddleware {
  /**
   * Middleware that only allows authenticated users to pass through to the next
   * handler.
   * @returns next if user exists in session
   * @returns 401 if user does not exist in session
   */
  authenticate: ControllerHandler = (req, res, next) => {
    // Check if user is authenticated
    try {
      const token = req.cookies.jwt
      req.user = decodeUserJWT(req)
    } catch (error) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in.' })
    }
    return next()
  }
}
