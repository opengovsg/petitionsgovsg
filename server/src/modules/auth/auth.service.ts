import { Post, PostStatus } from '~shared/types/base'
import { UserCannotViewPostError } from '../post/post.errors'

export class AuthService {
  /**
   * Check if a user is able to view the post
   * @param post post to be viewed
   * @param userId id of user
   * @returns true if user can view post
   */
  verifyUserCanViewPost = (post: Post, userId?: string): void => {
    // If post is public, anyone can view
    // If private or archived, must be logged in
    if (
      post.status === PostStatus.Open ||
      post.status === PostStatus.Draft ||
      userId
    )
      return
    throw new UserCannotViewPostError()
  }

  verifyPetitionOwner = (post: Post, userId?: string): boolean => {
    // If post is private or archived, only petition owner sees a personalised message
    if (userId) {
      // check that user id is the same as post user id
      const postUserId = post.hashedUserSgid
      return postUserId === userId
    }
    return false
  }
}
