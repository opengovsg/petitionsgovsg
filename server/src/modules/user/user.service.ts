import { User } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'
import { LoadPublicUserDto } from '~shared/types/api'
export class UserService {
  private User: ModelDef<User>

  constructor({ User }: { User: ModelDef<User> }) {
    this.User = User
  }

  loadUser = async (userId: number): Promise<LoadPublicUserDto> => {
    return this.User.findByPk(userId) as Promise<LoadPublicUserDto>
  }

  loadFullUser = async (userId: number): Promise<User> => {
    return this.User.findByPk(userId) as Promise<User>
  }

  loadUserBySgid = async (sgid: string): Promise<LoadPublicUserDto> => {
    return this.User.findOne({ where: { sgid: sgid } })
  }
}
