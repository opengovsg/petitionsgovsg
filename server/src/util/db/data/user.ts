import { UserAuthType } from '~shared/types/api'

export const mockUser = {
  id: 'sgid',
  fullname: 'John Doe',
  type: UserAuthType.Public,
}

export const mockOtherUser = {
  id: 'otherUser',
  fullname: 'Not John Doe',
  type: UserAuthType.Public,
}

export const mockAnonUser = {
  id: 'sgid',
  fullname: null,
  type: UserAuthType.Public,
}

export const mockUserJWT = {
  jwt: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InNnaWQiLCJmdWxsbmFtZSI6IkpvaG4gRG9lIiwidHlwZSI6InB1YmxpYyIsImlhdCI6MTY0NDIxMjcxNn0.SXDePYRMfpSEQaGX0YcloTkF_BvPSNiOkhcAoQ6W7NM',
}

export const mockOtherUserJWT = {
  jwt: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6Im90aGVyVXNlciIsImZ1bGxuYW1lIjoiTm90IEpvaG4gRG9lIiwidHlwZSI6InB1YmxpYyIsImlhdCI6MTY0NDQ2Mzk2MX0.xFLcMyOeg1B9h66KGRLMvNglmg2TUCUF9q1Zqnd9bwE',
}

export const mockAnonUserJWT = {
  jwt: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InNnaWQiLCJ0eXBlIjoicHVibGljIiwiZnVsbG5hbWUiOm51bGwsImlhdCI6MTY0NDQ2NTIyM30.6yuMIvcfFt0bwGHupMOeG-9SJtyJl4oPj_LPadyFLaY',
}
