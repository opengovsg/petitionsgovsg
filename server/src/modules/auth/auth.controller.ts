import { StatusCodes } from 'http-status-codes'
import { isEmpty } from 'lodash'
import passport from 'passport'
import {
  callbackRedirectUnauthorisedURL,
  callbackRedirectURL,
} from '../../bootstrap/config/auth'
import { Message } from 'src/types/message-type'
import { ErrorDto, LoadPublicUserDto, UserAuthType } from '~shared/types/api'
import { createLogger } from '../../bootstrap/logging'
import { UserService } from '../../modules/user/user.service'
import { ControllerHandler } from '../../types/response-handler'

const logger = createLogger(module)

export class AuthController {
  private userService: Public<UserService>

  constructor({ userService }: { userService: Public<UserService> }) {
    this.userService = userService
  }

  /**
   * Fetch logged in user details after being authenticated.
   * @returns 200 with user details
   * @returns 500 if user id not found
   * @returns 500 if database error
   */
  loadUser: ControllerHandler<unknown, LoadPublicUserDto | ErrorDto> = async (
    req,
    res,
  ) => {
    const id = req.user?.id
    const type = req.user?.type
    if (!id) {
      logger.error({
        message: 'User not found after being authenticated',
        meta: {
          function: 'loadUser',
          userId: req.user?.id,
        },
      })
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }

    if (type === UserAuthType.Public) {
      try {
        const user = await this.userService.loadUser(id)
        if (user) {
          return res.status(StatusCodes.OK).json({ id: user.id })
        } else {
          logger.error({
            message: 'User not found',
            meta: {
              function: 'loadUser',
              userId: req.user?.id,
            },
          })
          return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
        }
      } catch (error) {
        logger.error({
          message: 'Database Error while loading public user',
          meta: {
            function: 'loadUser',
            userId: req.user?.id,
          },
          error,
        })
        return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
      }
    }

    if (type === UserAuthType.Public) {
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }
  }

  /**
   * Logout
   * @returns 200 if logged out
   */
  handleLogout: ControllerHandler = (req, res) => {
    if (!req.session || isEmpty(req.session)) {
      logger.error({
        message: 'Attempted to sign out without a session',
        meta: {
          function: 'handleLogout',
        },
      })
      return res.sendStatus(StatusCodes.BAD_REQUEST)
    }

    req.session.destroy((error) => {
      if (error) {
        logger.error({
          message: 'Failed to destroy session',
          meta: {
            action: 'handleLogout',
            function: 'handleLogout',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Sign out failed' })
      }

      // No error.
      res.clearCookie('connect.sid')
      return res.status(StatusCodes.OK).json({ message: 'Sign out successful' })
    })
  }

  handleSgidLogin: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { redirect: string }
  > = async (req, res, next) => {
    const { redirect } = req.query
    // store redirect to post in state
    if (redirect) {
      passport.authenticate('sgid-with-name', { state: redirect })(
        req,
        res,
        next,
      )
    } else {
      passport.authenticate('sgid-with-name')(req, res, next)
    }
  }
  /**
   * Verify otp received by the user
   * @params code
   * @params state
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   */
  handleSgidCallback: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { code: string; state: string | undefined }
  > = async (req, res, next) => {
    const { state } = req.query
    passport.authenticate(
      'sgid-with-name',
      {},
      (error, user, info: Message) => {
        if (error) {
          logger.error({
            message: 'Error while authenticating',
            meta: {
              function: 'handleSgidCallback',
            },
            error,
          })
          return res.redirect(callbackRedirectUnauthorisedURL)
        }
        if (!user) {
          logger.warn({
            message: info.message,
            meta: {
              function: 'handleSgidCallback',
            },
          })
          res.redirect(callbackRedirectUnauthorisedURL)
        }
        req.logIn(user, (error) => {
          if (error) {
            logger.error({
              message: 'Error while logging in',
              meta: {
                function: 'handleSgidCallback',
              },
              error,
            })
            return res.redirect(callbackRedirectUnauthorisedURL)
          }
          //
          /**
           * Regenerate session to mitigate session fixation
           * We regenerate the session upon logging in so an
           * anonymous session cookie cannot be used
           */
          const passportSession = req.session.passport
          req.session.regenerate(function (error) {
            if (error) {
              logger.error({
                message: 'Error while regenerating session',
                meta: {
                  function: 'handleSgidCallback',
                },
                error,
              })
              return res.redirect(callbackRedirectUnauthorisedURL)
            }
            //req.session.passport is now undefined
            req.session.passport = passportSession
            req.session.save(function (error) {
              if (error) {
                logger.error({
                  message: 'Error while saving regenerated session',
                  meta: {
                    function: 'handleSgidCallback',
                  },
                  error,
                })
                return res.redirect(callbackRedirectUnauthorisedURL)
              }
              return res.redirect(callbackRedirectURL(state))
            })
          })
        })
      },
    )(req, res, next)
  }
}
