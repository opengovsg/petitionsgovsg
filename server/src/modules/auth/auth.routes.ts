import express from 'express'
import { AuthController } from './auth.controller'
import { AuthMiddleware } from './auth.middleware'

export const routeAuth = ({
  controller,
  authMiddleware,
}: {
  controller: AuthController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()

  /**
   * Fetch logged in user details
   * @route   GET /api/auth
   * @returns 200 with user details
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.get('/', authMiddleware.authenticate, controller.loadUser)

  /**
   * Logout
   * @route   POST /api/auth/logout
   * @returns 200 if logged out
   * @access  private
   */
  router.post('/logout', controller.handleLogout)

  /**
   * Sgid Login
   * @route   POST /api/auth/sgid/login
   * @returns 302 with sgid oidc link
   * @access  private
   */
  router.get('/sgid/login', controller.handleSgidLogin)

  /**
   * Sgid Callback
   * @route   GET /api/auth/callback
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   * @access  private
   */
  router.get('/callback', controller.handleSgidCallback)

  return router
}
