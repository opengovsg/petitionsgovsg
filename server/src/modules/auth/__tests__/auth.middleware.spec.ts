import express from 'express'
import { StatusCodes } from 'http-status-codes'
import supertest from 'supertest'
import { ControllerHandler } from '@/types/response-handler'
import { mockUserJWT } from '../../../util/db/data/user'
import { AuthMiddleware } from '../auth.middleware'

describe('auth.middleware', () => {
  // Set up auth middleware to inject user
  let jwt: { jwt: string } | undefined
  const authMiddleware = new AuthMiddleware()

  // Middleware to attach cookies
  const middleware: ControllerHandler = (req, _res, next) => {
    req['cookies'] = jwt
    next()
  }

  const authPath = '/auth'

  const app = express()
  app.use(express.json())
  app.use(middleware)
  app.get(authPath, authMiddleware.authenticate)
  app.use(function (req, res) {
    res.end()
  })
  const request = supertest(app)

  beforeEach(async () => {
    jest.resetAllMocks()
  })

  describe('authenticate', () => {
    it('should pass on to the next handler if authenticated', async () => {
      jwt = mockUserJWT

      const response = await request.get(authPath)

      expect(response.status).toEqual(StatusCodes.OK)
    })

    it('returns UNAUTHORIZED on no user', async () => {
      jwt = undefined

      const response = await request.get(authPath)

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toMatchSnapshot()
    })
  })
})
