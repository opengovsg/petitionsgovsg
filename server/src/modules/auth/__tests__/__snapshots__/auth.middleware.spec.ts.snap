// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth.middleware authenticate returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in.",
}
`;
